Sprint8 task2 stock management

iAcay opened the pull request:
Hey! The code which is below was implemented in order to meet the acceptance criterias of the second task in sprint no. 8 :)

AC in this task:
-- Each reward has a number of available items an employee could buy. For an online reward, the available items equal to the number of associated online codes
-- Admin can manage online codes and assign them to a reward
-- Admin can import online codes and associate them with a reward
-- The online code should be present in the online delivery e-mail
-- After delivering an order the number of available items should decrease
-- A reward with an online delivery method that has 0 available online codes shouldn't be visible to the user
-- An online reward should be delivered automatically
-- Online code can be used only once


https://user-images.githubusercontent.com/96024046/187306209-906cce85-8b3b-4564-99d5-01dc96d484bf.mov

iAcay committed:
feat: add online codes

-- create table online_codes
-- implement online_code model and online_codes controller
-- create view templates for online_codes crud actions
-- add link to online_codes to admin dashboard and navbar
-- add online code to online delivery e-mail
-- implement number_of_available_items method for checking amount of usable online_codes
-- add possibility for admin to import online codes from csv file
-- adjust tests for introducing online codes

iAcay committed:
feat: add service object for creating an order

-- replace OrderForm by CreateOrderService
-- reward with online delivery method is delivered automatically

pstrzalk commented on app/controllers/admin_users/online_codes_controller.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+
+    def update
+      if online_code.update(online_code_params)
```
can it be always updated or are there any cases when it shouldn't?

iAcay replied:
Hey, 
first of all thanks a lot for your CR :) 
At now we don't have any constraints with updating online codes. Thus in my opinion online code can be always updated.

pstrzalk replied:
Sounds good, thanks!

pstrzalk commented on app/models/reward.rb:
```diff
@@ -15,4 +16,14 @@ class Reward < ApplicationRecord
+  end
+
+  def available_for_purchase?
+    return true if delivery_method_post_delivery? || number_of_available_items.positive?
```
`number_of_available_items` may return a string, which doesn't have `.positive?` method. It would cause an error

iAcay replied:
Understood :) Honestly I thought about it when I was implementing these methods. I summarized that this kind of error would not occur because `number_of_available_items` returns string only when we have post delivery. The `available_for_purchase?` method first checks if reward has post delivery and (if it's true) returns true. So I assumed that there is no option for calling `.positive? method on string.

I also see the argument that a method should return only one type of value 👍  Currently acceptance criterias don't mention about manner of couting the number of available items when delivery method is post delivery. Thus I decided to return `0` insted of `undetermined` string. At now `available_for_purchase?` method returns true directly when a reward has post delivery so it shouldn't be a problem. :)

pstrzalk replied:
The construct which you had created would work as you intended but lacks readability as it's hard to reveal your intention. It's better when a reader doesn't have to double read a piece of code to get how it works and why. A single type of returned value makes its usage much cleaner and having less conditional branching.

You can probably reduce this method to
```ruby
def available_for_purchase?
  delivery_method_post_delivery? || number_of_available_items.positive?
end
```

:)

iAcay replied:
Understood! Thanks :) Of course I changed syntax according to your suggestion 👍

pstrzalk commented on app/controllers/orders_controller.rb:
```diff
@@ -37,7 +36,7 @@ def address_params
   end
 
-  def order_form_params
+  def create_order_params
```
it might be good to return `{}` from address_params if there is no `params[:address]`. Currently it returns `nil` which would make `order_params.merge(address_params)` throw a `TypeError`

iAcay replied:
Understood! :) I changed it 👍

pstrzalk commented on app/models/reward.rb:
```diff
@@ -15,4 +16,14 @@ class Reward < ApplicationRecord
   end
+
+  def number_of_available_items
+    delivery_method_online? ? online_codes.where(status: :not_used).count : 'undetermined'
```
it's in general a code smell if a method returns more than one type of value. In here - a number or a string. It's common however, to use a value of some type or `nil`. 


Since `number_of_available_items` is not used outside of this model, maybe it would be easier to just merge its logic into `available_for_purchase?`? 
At minimum, it can be made `private`

iAcay replied:
I have changed it --> currently method returns only one type of value 👍 

I'm a little bit confused about `number_of_available_items` method. Of course this method currently is not used outside of a reward model but one of the acceptance criterias says that _"Each reward has a number of available items"._ I understand above criteria in this way -> for each reward we can check the number of available items by calling a method. For this reason I decided to divide and keep both methods public. What do you think about it? :)

pstrzalk replied:
For the acceptance criteria - it doesn't say it has to be publicly accessible :) 

Personally, I wouldn't keep a public, unused, method. But if you think it's something that will be an important interface, used in the future, then it's 100% up to you.

iAcay replied:
Good point :D I moved the `number_of_available_items` method to private 👍  Thanks!

pstrzalk commented on app/views/admin_users/online_codes/new_import_from_csv.html.erb:
```diff
@@ -0,0 +1,15 @@
+      <h2>Import online codes from CSV file</h2>
+    </div>
+    <div class="card-body">
+      <%= form_for online_code, url: import_from_csv_admin_users_online_codes_path do |f| %>
```
is this form really for `online_code`? Or is it a generic form with generic URL?

iAcay replied:
Unfortunately I am not pretty sure what do you mean, but I changed `form_for online_code` into `form_with`.

pstrzalk replied:
nice, that's exactly what I've aimed for. 
What I meant was - you had created a form which was intended to handle an OnlineCode resource. But in fact you needed a custom form for a custom action (with no context of any OnlineCode things).

iAcay replied:
:)

pstrzalk commented on app/services/online_codes_import_service.rb:
```diff
@@ -0,0 +1,32 @@
+require 'csv'
+
+class OnlineCodesImportService
```
It would be awesome if this class had some test coverage

iAcay replied:
I agree, but acceptance criterias in this task do not mention about tests so I didn't know if I should go further :) Of course I can try implementing tests for that feature :)

pstrzalk replied:
Acceptance criteria (given by a client) almost never mentions tests :) It's up to a developer to deliver them. 

In the case of a training course like this one, it's probably up to you whether you'd like to train that part of the job.

iAcay replied:
OK! In that case I will implement tests :)

pstrzalk commented on app/services/create_order_service.rb:
```diff
@@ -0,0 +1,61 @@
+class CreateOrderService
```
It would be awesome if this class had some test coverage

iAcay replied:
I agree, but acceptance criterias in this task do not mention about tests so I didn't know if I should go further :) Of course I can try implementing tests for that feature :)

pstrzalk replied:
As before, up to you what you'd like to train. Just please let me know so I can approve the PR at your desired state.

iAcay replied:
I will implement tests 👍 Thus please wait with approving the PR until the tests will be commited :)

pstrzalk replied:
👍

iAcay replied:
Thanks for your waiting :) I added tests for both service objects. Could you take a look at them?

pstrzalk commented on app/services/create_order_service.rb:
```diff
@@ -0,0 +1,61 @@
+
+  def send_email_to_employee
+    @order.status_delivered!
+    EmployeeMailer.delivery_confirmation_email(@order).deliver_now
```
this line should be outside of database transaction (it doesn't write anything to DB, yet it is a lengthy operation). You may extract it to a new method and call separately

iAcay replied:
Understood 👍  I extracted sending email method from transaction :)

pstrzalk replied:
Great, thanks

pstrzalk reviewed (commented):
Hey,
It's quite a big PR so I didn't go into much details but I commented on some things that stood out at the first read.

Please let me know what you think - especially if I should elaborate on any of the comments of if you'd like to discuss any of them.

iAcay committed:
fix: add return {} option when params[:address] are nil

iAcay committed:
fix: only one type of value is returned from method

iAcay committed:
fix: extract sending email method from transaction

iAcay committed:
refactor: change form_for into form_with

pstrzalk commented on app/controllers/admin_users/online_codes_controller.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+
+    def new_import_from_csv
+      render :new_import_from_csv, locals: { online_code: OnlineCode.new }
```
You don't need `online_code` local since you've updated the form

iAcay replied:
That's right 📲  Local has been removed :) Thanks again! 👍

iAcay committed:
fix: remove unnecessary local variable

iAcay committed:
refactor: move number_of_available_items method to private

iAcay committed:
tests: check that the service objects are working properly

-- tests for CreateOrderService class
-- tests for OnlineCodesImportService class

pstrzalk commented on spec/services/online_codes_import_service_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      create(:reward, title: 'RewardTest1')
+      create(:reward, title: 'RewardTest2')
+      params = { file: fixture_file_upload('./spec/fixtures/online_codes/valid_online_codes_import.csv') }
+      import_online_codes = described_class.new(params)
```
Hint: if you call such a variable `service`, it is always obvious what it is. `import_online_codes` doesn't match service's class name, so it may be not readable at first glance.

iAcay replied:
Understood! I have changed it :) Thanks!

pstrzalk commented on spec/services/online_codes_import_service_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      params = { file: fixture_file_upload('./spec/fixtures/online_codes/valid_online_codes_import.csv') }
+      import_online_codes = described_class.new(params)
+
+      expect(OnlineCode.count).to eq 0
+      expect(import_online_codes.import).to be true
+      expect(OnlineCode.count).to eq 2
```
it won't always be true. There are some gimmics when it comes to running tests in real pipelines - database is not always cleared down to no resources before each test.

Also, testing for count being 0 is pointless from the perspective of the production code, as it tests the test itself

The test should look more or less like this (written from head, forgive typos please)
```ruby

# initiate the variable here so that it is overwritten in the block below, 
# otherwise `result = ...` in the block wouldn't be visible outside of the block
result = nil 

# pass block to `expect` to be able to measure changes within execution
expect { result = import_online_codes.import }.to change(OnlineCode, :count).by(2)
expect(result).to be true
```

iAcay replied:
Got it 👍  I have changed it according to your suggestion :) Thanks!

pstrzalk commented on spec/services/create_order_service_spec.rb:
```diff
@@ -0,0 +1,94 @@
+      address_params = attributes_for(:address)
+      create_order_service = described_class.new(order_params.merge(address_params))
+
+      expect(Address.count).to eq 1
+      expect(employee.address.street).to eq first_address.street
+      expect(employee.address.postcode).to eq first_address.postcode
+      expect(employee.address.city).to eq first_address.city
```
This part of the scenario tests the test itself. I would have faith in the framework and remove those lines.

iAcay replied:
It has been removed 👍

pstrzalk commented on spec/services/create_order_service_spec.rb:
```diff
@@ -0,0 +1,94 @@
+      expect { create_order_service.call }.not_to change(Address, :count)
+
+      employee.reload
+      expect(employee.address.street).not_to eq first_address.street
+      expect(employee.address.postcode).not_to eq first_address.postcode
+      expect(employee.address.city).not_to eq first_address.city
+      expect(employee.address.street).to eq address_params[:street]
+      expect(employee.address.postcode).to eq address_params[:postcode]
+      expect(employee.address.city).to eq address_params[:city]
```
if you explicitly define the values of `address_params` in the beginning of this scenario, then you can just test for those values in here. For sure you don't have to test for them being different than something and then for being equal to smth else.

Tests become much more readable when there are explicit values involved - even if they are set to "foo" or 7

so this block may be then changed to something like
```ruby
expect(employee.address).to have_attributes(
  street: 'Downing Street',
  postcode: 'AB00111',
  city: 'London'
)
```

iAcay replied:
Thanks for explanation 👍 Of course I used your suggestion and changed code :)

pstrzalk reviewed (commented):
Thank you for adding tests for these two service.

I've commented on them, mostly in the fashion of "here are some hints - use them if you'd like", as you've already accomplished the main goal of having at least some test coverage :)

Please write a comment when you think all is ready (even if you'd like to leave the test as they are). I will take the last look and hopefully I can just approve :)

Thanks!

iAcay committed:
tests: refactor syntax according to comments from cr

iAcay commented:
> Thank you for adding tests for these two service.
> 
> I've commented on them, mostly in the fashion of "here are some hints - use them if you'd like", as you've already accomplished the main goal of having at least some test coverage :)
> 
> Please write a comment when you think all is ready (even if you'd like to leave the test as they are). I will take the last look and hopefully I can just approve :)
> 
> Thanks!

Thanks a lot for your hints! I will remember them for the future! 

I think that all is ready 👍

pstrzalk commented on app/views/rewards/index.html.erb:
```diff
@@ -50,7 +50,11 @@
           <figcaption class="block-footer text-end">
             <%= link_to 'Show', reward_path(reward), class: 'btn btn-outline-primary' %>
-            <%= link_to 'Buy', new_order_path(reward: reward), class: 'btn btn-outline-success' %>
+            <% if reward.available_for_purchase? %>
```
One of the acceptance criteria says "A reward with an online delivery method that has 0 available online codes shouldn't be visible to the user"

I'm not too familiar with the entire app, but I would assume that it means that `Show` is also restricted.

A second point - controller actions should also restrict access if this condition is not met. So `new_order_path(reward: reward)` should reject request (or redirect somewhere) unless `reward.available_for_purchase?`. 

Please let me know if I'm not getting the context correctly.

iAcay replied:
Honestly, I wasn't sure how to understand this ac. Your argument is right so I implemented restrictions for `Show` action when reward is not available for purchase. These restrictions were added to index.html.erb file and also to rewards_controller.rb :)

About second point, CreateOrderService first checks if reward is for sale (it uses `available_for_purchase?` method). If reward is not then it returns false from entire `call` method and prints error _"Reward is not available now."_ Please let me know if it's enough :) Currently the check looks:

````
  def reward_for_sale?
    return true if @reward.available_for_purchase?

    @errors << 'Reward is not available now.'
    false
  end
````

pstrzalk replied:
I think that in this shape of the project the service logic is enough.
Thank you for the updates!

iAcay replied:
Thanks a lot! 👍

iAcay committed:
feat: show action is restricted when online reward hasn't online code

pstrzalk reviewed (approved):
Thank you for all the updates!
Good luck with the next task :)

iAcay commented:
Thank you a lot for the entire review! :)
